packages/SwingSet/src/kernel/vatManager/vat-warehouse.js

```diff
@@ -0,0 +1,275 @@
+
+/**
+ * @param { ReturnType<typeof import('../state/kernelKeeper').default> } kernelKeeper
+ * @param { ReturnType<typeof import('../loadVat').makeVatLoader> } vatLoader
+ * @param {{ maxVatsOnline?: number }=} policyOptions
+ *
+ * @typedef {(syscall: VatSyscallObject) => ['error', string] | ['ok', null] | ['ok', Capdata]} VatSyscallHandler
+ * @typedef {{ body: string, slots: unknown[] }} Capdata
+ * @typedef { [unknown, ...unknown[]] } Tagged
+ * @typedef { { moduleFormat: string }} Bundle
+ */
+export function makeVatWarehouse(kernelKeeper, vatLoader, policyOptions) {
+  const { maxVatsOnline = 50 } = policyOptions || {};
+  // console.debug('makeVatWarehouse', { policyOptions });
+
+  /**
+   * @typedef {{
+   *   manager: VatManager,
+   *   enablePipelining: boolean,
+   *   options: { name?: string, description?: string },
+   * }} VatInfo
+   * @typedef { ReturnType<typeof import('../vatTranslator').makeVatTranslators> } VatTranslators
+   */
+  const ephemeral = {
+    /** @type {Map<string, VatInfo> } key is vatID */
+    vats: new Map(),
+  };
+
+  /** @type {Map<string, VatTranslators> } */
+  const xlate = new Map();
+  /** @param { string } vatID */
+  function provideTranslators(vatID) {
+    let translators = xlate.get(vatID);
+    if (!translators) {
+      // NOTE: makeVatTranslators assumes
+      // vatKeeper for this vatID is available.
+      translators = makeVatTranslators(vatID, kernelKeeper);
+      xlate.set(vatID, translators);
+    }
+    return translators;
+  }
+
+  /**
+   * @param {string} vatID
+   * @param {boolean} recreate
+   * @returns { Promise<VatInfo> }
+   */
+  async function ensureVatOnline(vatID, recreate) {
+    const info = ephemeral.vats.get(vatID);
+    if (info) return info;
+
+    const vatKeeper =
+      kernelKeeper.getVatKeeper(vatID) || kernelKeeper.allocateVatKeeper(vatID);
+    const { source, options } = vatKeeper.getSourceAndOptions();
+
+    const translators = provideTranslators(vatID);
+
+    const chooseLoader = () => {
+      if (recreate) {
+        const isDynamic = kernelKeeper.getDynamicVats().includes(vatID);
+        if (isDynamic) {
+          return vatLoader.recreateDynamicVat;
+        } else {
+          return vatLoader.recreateStaticVat;
+        }
+      } else {
+        return vatLoader.createVatDynamically;
+      }
+    };
+    // console.log('provide: creating from bundle', vatID);
+    const manager = await chooseLoader()(vatID, source, translators, options);
+    assert(manager, `no vat manager; kernel panic?`);
+
+    // TODO(3218): persist this option; avoid spinning up a vat that isn't pipelined
+    const { enablePipelining = false } = options;
+
+    // TODO: load from snapshot
+    await manager.replayTranscript();
+    const result = {
+      manager,
+      translators,
+      enablePipelining,
+      options,
+    };
+    ephemeral.vats.set(vatID, result);
+    // eslint-disable-next-line no-use-before-define
+    applyAvailabilityPolicy(vatID);
```

Comment:
This call to `applyAvailabilityPolicy` is lacking an `await`, which means any evictions it does will happen in parallel with whatever the caller does next. Since `evict` might cause DB state changes (snapshot ID, transcript start offset), it's probably not a good idea to allow those to occur at random times.

Let's add an `await` here. We still need to think through the "what state changes get committed in which cranks" question, but at least doing an `await` will get us behavior that's consistent, if not necessarily correct.